File: m/rls/firea.r
file	firea - read/write file
#undef __STDC__				; need unix things
include <stdio.h>
include rid:rider
include rid:fidef
include rid:stdef
If Wdw || Wnt || Msc
#undef putchar
include	<io.h>
include	rid:cldef
End

;	fi_rea/fi_wri read 1 element of n bytes. Thus, they cannot report
;	the size of partial read/write (use fi_ipt/fi_opt). fi_rea/fi_wri
;	may be significantly faster than fi_ipt/fi_opt.

code -  fi_rea - read file

  func	fi_rea
	fil : * FILE
	buf : * void
	cnt : size_t
	()  : int 			; fine/fail
  is	fread (buf, cnt, 1, fil)	; read it
	reply that eq 1			; success
  end

code -  fi_wri - write file

  func	fi_wri
	fil : * FILE
	buf : * void
	cnt : size_t
	()  : int 			; fine/fail
  is	cnt = st_len(buf) if cnt eq ~(1); mumble mumble
If Wdw
	tmp : * char
	if cl_tty (fil)			; is console file
	   tmp = buf			; avoid casting
	   putchar (*tmp++) while cnt--	; ugh
	.. fine				;
End					;
	fwrite (buf, cnt, 1, fil)	; write it
	reply that eq 1			; wrote it all
  end


File: m/rls/rlb/firea.c
/* file -  firea - read/write file */
#undef __STDC__
#include <stdio.h>
#include "m:\rid\rider.h"
#include "m:\rid\fidef.h"
#include "m:\rid\stdef.h"
#if Wdw || Wnt || Msc
#undef putchar
#include <io.h>
#include "m:\rid\cldef.h"
#else 
#include "m:\rid\dsdef.h"
#endif 
/* code -  - fi_rea - read file */
int fi_rea(
FILE *fil ,
void *buf ,
size_t cnt )
{ 
  return ( (fread (buf, cnt, 1, fil)) == 1);
} 
/* code -  - fi_wri - write file */
int fi_wri(
FILE *fil ,
void *buf ,
size_t cnt )
{ if ( cnt == (~1)) {cnt = st_len(buf) ;}
#if Wdw
  char *tmp ;
  if ( cl_tty (fil)) {
    tmp = buf;
    while ( cnt--) {putchar (*tmp++) ;}
    return 1; }
#endif 
  return ( (fwrite (buf, cnt, 1, fil)) == 1);
} 
#if Dos
/* code -  fi_drd - direct read file */
int fi_drd(
FILE *fil ,
void *buf ,
size_t cnt )
{ 
  dsTreg reg ;
  dsTseg seg ;
  DS=SEG(buf), DX=OFF(buf);
  BX = fileno (fil), CX=cnt;
  ds_s21 (0x3f00);
  return ( !CF && AX == cnt);
} 
/* code -  fi_dwr - direct write file */
int fi_dwr(
FILE *fil ,
void *buf ,
size_t cnt )
{ 
  dsTreg reg ;
  dsTseg seg ;
  if( cnt == 0)return 1;
  DS=SEG(buf), DX=OFF(buf);
  BX = fileno (fil), CX=cnt;
  ds_s21 (0x4000);
  return ( !CF && AX == cnt);
} 
#endif 
